server/src/metadata.rs

```diff
@@ -0,0 +1,1310 @@
+            field_columns.push(fallback_field_column);
+        }
+
+        // Retrieve the hashes and compute the keys;
+        let mut select_statement = connection.prepare(&query_hashes)?;
+        let mut rows = select_statement.query([])?;
+
+        let mut keys = vec![];
+        while let Some(row) = rows.next()? {
+            // SQLite use signed integers https://www.sqlite.org/datatype3.html.
+            let signed_tag_hash = row.get::<usize, i64>(0)?;
+            let tag_hash = u64::from_ne_bytes(signed_tag_hash.to_ne_bytes());
+
+            for field_column in &field_columns {
+                keys.push(tag_hash | field_column);
+            }
+        }
+        Ok(keys)
+    }
+
+    /// If they do not already exist, create the tables used for table and model
+    /// table metadata. A "table_metadata" that can persist tables is created, a
+    /// "model_table_metadata" table that can persist model tables is created,
+    /// and a "columns" table that can save the index of field columns in
+    /// specific model tables is created. If the tables exist or were created,
+    /// return [`Ok`], otherwise return [`rusqlite::Error`].
+    fn create_metadata_database_tables(&self) -> Result<()> {
+        let connection = Connection::open(&self.metadata_database_path)?;
+
+        // Create the table_metadata SQLite table if it does not exist.
+        connection.execute(
+            "CREATE TABLE IF NOT EXISTS table_metadata (
+                table_name TEXT PRIMARY KEY
+        ) STRICT",
+            (),
+        )?;
+
+        // Create the model_table_metadata SQLite table if it does not exist.
+        connection.execute(
+            "CREATE TABLE IF NOT EXISTS model_table_metadata (
+                table_name TEXT PRIMARY KEY,
+                schema BLOB NOT NULL,
+                timestamp_column_index INTEGER NOT NULL,
+                tag_column_indices BLOB NOT NULL
+        ) STRICT",
+            (),
+        )?;
+
+        // Create the model_table_field_columns SQLite table if it does not
+        // exist. Note that column_index will only use a maximum of 10 bits.
+        connection.execute(
+            "CREATE TABLE IF NOT EXISTS model_table_field_columns (
+                table_name TEXT NOT NULL,
+                column_name TEXT NOT NULL,
+                column_index INTEGER NOT NULL,
+                PRIMARY KEY (table_name, column_name)
+        ) STRICT",
+            (),
+        )?;
+
+        Ok(())
+    }
+
+    /// Return [`true`] if `path` is a data folder, otherwise [`false`].
+    fn is_path_a_data_folder(path: &Path) -> bool {
+        if let Ok(files_and_folders) = fs::read_dir(path) {
+            files_and_folders.count() == 0 || path.join(METADATA_SQLITE_NAME).exists()
+        } else {
+            false
+        }
+    }
+}
+
+/// Metadata required to ingest data into a model table and query a model table.
+#[derive(Debug, Clone)]
+pub struct ModelTableMetadata {
```

Comment:
I have moved `ModelTableMetadata` and the associated tests to a separate sub-module. Although, it does not make much of different as `model_table_medata.rs` is only 291 lines of code and tests while `mod.rs` is 1071 lines of code and tests. So it may have been simpler to keep it all in one file with 1362 lines of code, however, now it is simpler to add new sub-modules if needed in the future.